lib/parser/message.cc

```diff
@@ -33,29 +33,92 @@ std::ostream &operator<<(std::ostream &o, const MessageFixedText &t) {
   return o;
 }
 
+static bool NeedsSpecialFormatting(const char *p) {
+  while (*p != '\0') {
+    if (*p++ == '%') {
+      if (*p == 'S' || *p == 'B') {
+        return true;
+      } else if (*p != '%' && *p != 's' && *p != 'd' &&
+          !((*p == 'z' || *p == 'j') && (p[1] == 'd' || p[1] == 'u'))) {
+        return false;
```

Comment:
If we hit this case, shouldn't we keep going to see if there is %S or %B after this code we don't handle? Then we could assert rather than outputting a bad message.

Reply:
Done.